src/detect-http2.c

```diff
@@ -0,0 +1,286 @@
+ *
+ * \retval 0 no match
+ * \retval 1 match
+ */
+static int DetectHTTP2frametypeMatch(DetectEngineThreadCtx *det_ctx,
+                               Flow *f, uint8_t flags, void *state, void *txv, const Signature *s,
+                               const SigMatchCtx *ctx)
+
+{
+    uint8_t *detect = (uint8_t *)ctx;
+
+    int value = rs_http2_tx_get_frametype(txv, flags);
+    if (value < 0) {
+        //no value, no match
+        return 0;
+    }
+    return *detect == value;
+}
+
+static int DetectHTTP2FuncParseFrameType(const char *str, uint8_t *ft)
+{
+    // first parse numeric value
+    if (ByteExtractStringUint8(ft, 10, strlen(str), str) >= 0) {
+        return 1;
+    }
+
+    // it it failed so far, parse string value from enumeration
+    int r = rs_http2_parse_frametype(str);
+    if (r >= 0) {
+        *ft = r;
+        return 1;
+    }
+
+    return 0;
+}
+
+/**
+ * \brief this function is used to attach the parsed http2.frametype data into the current signature
+ *
+ * \param de_ctx pointer to the Detection Engine Context
+ * \param s pointer to the Current Signature
+ * \param str pointer to the user provided http2.frametype options
+ *
+ * \retval 0 on Success
+ * \retval -1 on Failure
+ */
+static int DetectHTTP2frametypeSetup (DetectEngineCtx *de_ctx, Signature *s, const char *str)
+{
+    uint8_t frame_type;
+
+    if (DetectSignatureSetAppProto(s, ALPROTO_HTTP2) != 0)
+        return -1;
+
+    if (!DetectHTTP2FuncParseFrameType(str, &frame_type)) {
+        SCLogError(SC_ERR_INVALID_SIGNATURE,
+                   "Invalid argument \"%s\" supplied to http2.frametype keyword.", str);
+        return -1;
+    }
+
+    uint8_t *http2ft = SCCalloc(1, sizeof(uint8_t));
+    if (http2ft == NULL)
+        return -1;
+    *http2ft = frame_type;
+
+    SigMatch *sm = SigMatchAlloc();
+    if (sm == NULL) {
+        DetectHTTP2frametypeFree(http2ft);
+        return -1;
+    }
+
+    sm->type = DETECT_HTTP2_FRAMETYPE;
+    sm->ctx = (SigMatchCtx *)http2ft;
+
+    SigMatchAppendSMToList(s, sm, g_http2_match_buffer_id);
+
+    return 0;
+}
+
+/**
+ * \brief this function will free memory associated with uint8_t
+ *
+ * \param ptr pointer to uint8_t
+ */
+void DetectHTTP2frametypeFree(void *ptr)
+{
+    SCFree(ptr);
+}
+
+/**
+ * \brief This function is used to match HTTP2 error code rule option on a transaction with those passed via http2.errorcode:
+ *
+ * \retval 0 no match
+ * \retval 1 match
+ */
+static int DetectHTTP2errorcodeMatch(DetectEngineThreadCtx *det_ctx,
+                               Flow *f, uint8_t flags, void *state, void *txv, const Signature *s,
+                               const SigMatchCtx *ctx)
+
+{
+    uint32_t *detect = (uint32_t *)ctx;
+
+    int value = rs_http2_tx_get_errorcode(txv, flags);
+    if (value < 0) {
+        //no value, no match
+        return 0;
+    }
+    return *detect == (uint32_t) value;
```

Comment:
I guess we should improve this...
Any thoughts ?

Reply:
Seems like premature optimization to me.

Reply:
So, just leave a TODO comment for now ?

Reply:
It's likely we'll want things like negation or ranges here too

Reply:
Ok for negation, but not sure for ranges as this field is an enumeration

Reply:
And do not we have generic negation for whatever keyword (returning the negation of the Match function) ?